controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+public:
+   void Init( UART_Regs *r, int baud )
+   {
+      reg = r;
+
+      // Set baud rate register
+      reg->baud = CPU_FREQ / baud;
+
+      // Enable the UART and receive interrupts
+      reg->ctrl[0] = 0x002D;
+   }
+
+   void ISR( void ){
+
+      // Check for over run error and framing errors.
+      // Clear those errors if they're set to avoid
+      // further interrupts from them.
+      if( reg->status & 0x000A )
+         reg->intClear = 0x000A;
+
+      // See if we received a new byte
+      if( reg->status & 0x0020 )
+         rxDat.Put( reg->rxDat );
+
+      // Check for transmit data register empty
+      if( reg->status & reg->ctrl[0] & 0x0080 )
+      {
+         int ch = txDat.Get();
+
+         // If there's nothing left in the transmit buffer, 
+         // just disable further transmit interrupts.
+         if( ch < 0 )
+            reg->ctrl[0] &= ~0x0080;
+
+         // Otherwise, Send the next byte
+         else
+            reg->txDat = ch;
+      }
+   }
+
+   uint16_t read( uint8_t *buf, uint16_t len ){
+
+      for( uint16_t i=0; i<len; i++ ) {
+         int ch = rxDat.Get();
+         if( ch < 0 ) return i;
+         *buf++ = ch;
+      }
+      return len;
+   }
+
+   uint16_t write( const uint8_t *buf, uint16_t len) {
+
+      uint16_t i;
+      for( i=0; i<len; i++ ){
+         if( !txDat.Put( *buf++ ) )
+            break;
+      }
+
+      // Enable the tx interrupt
+      reg->ctrl[0] |= 0x0080;
+
+      return i;
+   }
+
+   int RxFull( void ){
+      return rxDat.FullCt();
+   }
+
+   int TxFree( void ){
+      return txDat.FreeCt();
+   }
+};
+
+static UART rpUART;
+
+// The UART that talks to the rPi uses the following pins:
+//    PB10 - TX
+//    PB11 - RX
+//    PB13 - RTS
+//    PB14 - CTS
+//
+// These pins are connected to UART3
+static void InitUARTs( void )
+{
+// NOTE - The UART functionality hasn't been tested due to lack of hardware!
+//        Need to do that as soon as the boards are available.
+   EnableClock( UART3_BASE );
+
+   GPIO_PinAltFunc( GPIO_B_BASE, 10, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 11, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 13, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 14, 7 );
+
+   rpUART.Init( (UART_Regs*)UART3_BASE, 115200 );
+
+   EnableInterrupt( INT_VEC_UART3, 3 );
+}
+
+static void UART3_ISR( void ){
+   rpUART.ISR();
+}
+
+uint16_t HalApi::serialRead(char *buf, uint16_t len) {
+   return rpUART.read( (uint8_t *)buf, len );
+}
+
+uint16_t HalApi::serialBytesAvailableForRead() {
+   return rpUART.RxFull();
+}
+
+uint16_t HalApi::serialWrite(const char *buf, uint16_t len) {
+   return rpUART.write( (const uint8_t *)buf, len);
+  return 0;
+}
+
+uint16_t HalApi::serialBytesAvailableForWrite() {
+   return rpUART.TxFree();
+}
+
+/******************************************************************
+ * Watchdog timer
+ *****************************************************************/
+void HalApi::watchdog_init() {
+}
+
+void HalApi::watchdog_handler() {
```

Comment:
It doesn't seem correct for this to be a no-op, am I missing something?

Reply:
No, just not finished yet.
I'll add watchdog timer support.

Reply:
OK, done and tested.